src/main/java/org/opensearch/knn/index/mapper/ModelFieldMapper.java

```diff
@@ -31,43 +53,109 @@ public class ModelFieldMapper extends KNNVectorFieldMapper {
         boolean stored,
         boolean hasDocValues,
         ModelDao modelDao,
-        String modelId,
         Version indexCreatedVersion
     ) {
-        super(simpleName, mappedFieldType, multiFields, copyTo, ignoreMalformed, stored, hasDocValues, indexCreatedVersion);
-
-        this.modelId = modelId;
+        super(simpleName, mappedFieldType, multiFields, copyTo, ignoreMalformed, stored, hasDocValues, indexCreatedVersion, null);
+        ANNConfig annConfig = mappedFieldType.getAnnConfig();
+        modelId = annConfig.getModelId().orElseThrow(() -> new IllegalArgumentException("KNN method context cannot be empty"));
         this.modelDao = modelDao;
 
+        this.spaceType = new AtomicReference<>(null);
+        this.methodComponentContext = new AtomicReference<>(null);
+        this.dimension = new AtomicInteger(UNSET_MODEL_DIMENSION_IDENTIFIER);
+        this.vectorDataType = new AtomicReference<>(null);
+        this.perDimensionProcessor = new AtomicReference<>(null);
+        this.perDimensionValidator = new AtomicReference<>(null);
+        this.vectorValidator = new AtomicReference<>(null);
+
         this.fieldType = new FieldType(KNNVectorFieldMapper.Defaults.FIELD_TYPE);
         this.fieldType.putAttribute(MODEL_ID, modelId);
         this.fieldType.freeze();
     }
 
     @Override
-    protected void parseCreateField(ParseContext context) throws IOException {
+    protected void validatePreparse() {
+        super.validatePreparse();
         // For the model field mapper, we cannot validate the model during index creation due to
         // an issue with reading cluster state during mapper creation. So, we need to validate the
-        // model when ingestion starts.
+        // model when ingestion starts. We do this as lazily as we can
         ModelMetadata modelMetadata = this.modelDao.getMetadata(modelId);
 
         if (!ModelUtil.isModelCreated(modelMetadata)) {
             throw new IllegalStateException(
                 String.format(
                     "Model \"%s\" from %s's mapping is not created. Because the \"%s\" parameter is not updatable, this index will need to be recreated with a valid model.",
                     modelId,
-                    context.mapperService().index().getName(),
+                    simpleName(),
                     MODEL_ID
                 )
             );
         }
 
-        parseCreateField(
-            context,
-            modelMetadata.getDimension(),
-            modelMetadata.getSpaceType(),
-            modelMetadata.getMethodComponentContext(),
-            modelMetadata.getVectorDataType()
-        );
+        maybeInitLazyVariables(modelMetadata);
+    }
+
+    private void maybeInitLazyVariables(ModelMetadata modelMetadata) {
```

Comment:
Shouldn't we just have a separate class and handle validation and processing there dynamically based on values in modelMetadata?

Reply:
Not sure I understand the suggestion

Reply:
Any specific reason of this style of atomic thing? Does this method called multiple time? Also, are these variables used multiple times and that is why we need to keep all these variable in instance level?

Can it be
```
private static final HashSet<VectorDataType, PerDimensionValidator> VALIDATOR = ...
validator.get(vectorDataType).validate();
```

Or 
```
private static final PerDimensionValidator perDimensionValidator;
perDimensionValidator.validate(vectorDataType);
```

Reply:
The problem with the model field mapper is that it relies on cluster state to store the model id. But during recovery, the cluster state is not available. So we cannot get the model metadata during field mapper creation. Related PR: https://github.com/opensearch-project/k-NN/pull/111. So, the idea is that we have to set these onces field parsing begins. The reason for atomics is to ensure they get set once safely.